contracts/fee/FeeHandler.sol

```diff
@@ -3,46 +3,176 @@
 
 // @title FeeHandler
-contract FeeHandler is ReentrancyGuard, RoleModule {
+contract FeeHandler is ReentrancyGuard, RoleModule, BasicMulticall {
+    using SafeERC20 for IERC20;
+
     DataStore public immutable dataStore;
     EventEmitter public immutable eventEmitter;
+    address public immutable vaultV1;
+    address public immutable gmx;
 
     constructor(
         RoleStore _roleStore,
         DataStore _dataStore,
-        EventEmitter _eventEmitter
+        EventEmitter _eventEmitter,
+        address _vaultV1,
+        address _gmx
     ) RoleModule(_roleStore) {
         dataStore = _dataStore;
         eventEmitter = _eventEmitter;
+        vaultV1 = _vaultV1;
+        gmx = _gmx;
+    }
+
+    // @dev withdraw fees from this contract
+    // @param token the token for which to claim fees
+    // @param amount the amount of fees to claim
+    function withdrawFees(address token, uint256 amount) external nonReentrant onlyFeeKeeper {
+        address receiver = dataStore.getAddress(Keys.FEE_RECEIVER);
+        IERC20(token).safeTransfer(receiver, amount);
     }
 
     // @dev claim fees from the specified markets
-    // @param markets the markets to claim fees from
-    // @param tokens the fee tokens to claim
-    function claimFees(
-        address[] memory markets,
-        address[] memory tokens
-    ) external nonReentrant onlyFeeKeeper {
-        if (markets.length != tokens.length) {
-            revert Errors.InvalidClaimFeesInput(markets.length, tokens.length);
+    // @param market the markets to claim fees from
+    // @param feeToken the fee tokens to claim
+    function claimFees(address market, address feeToken) external nonReentrant {
+        uint256 version;
+        uint256 buybackGmxFactor;
+        uint256 feeAmount;
+        if (market == address(0)) {
+            version = 1;
+            buybackGmxFactor = _getBuybackGmxFactor(version);
+            feeAmount = IVaultV1(vaultV1).withdrawFees(feeToken, address(this));
+            _setAvailableFeeAmounts(version, feeToken, feeAmount);
+        } else {
+            version = 2;
+            buybackGmxFactor = _getBuybackGmxFactor(version);
+            feeAmount = FeeUtils.claimFees(dataStore, eventEmitter, market, feeToken, address(this));
+            _setAvailableFeeAmounts(version, feeToken, feeAmount);
         }
+    }
 
-        address receiver = dataStore.getAddress(Keys.FEE_RECEIVER);
+    // @dev receive an amount in feeToken by depositing the batchSize amount of the buybackToken
+    // @param feeToken the token to receive with the fee amount calculated via an oracle price
+    // @param buybackToken the token to deposit in the amount of batchSize in return for fees
+    function buybackFees(address feeToken, address buybackToken) external nonReentrant {
+        uint256 batchSize = dataStore.getUint(Keys.buybackBatchAmountKey(buybackToken));
+        if (batchSize == 0) {
+            revert Errors.InvalidBuybackTokenInput(buybackToken);
```

Comment:
this error can be "InvalidBuybackToken" instead of "InvalidBuybackTokenInput"

Reply:
No problem, I've updated accordingly.

Reply:
I don't want to jump the gun, but I think this can be marked resolved now?